core/commands/dag/dag.go

```diff
@@ -241,3 +278,421 @@ var DagResolveCmd = &cmds.Command{
+				//
+				// // not ideal, but the pinning api takes only paths :(
+				// rp := path.NewResolvedPath(
+				// 	ipfspath.FromCid(c),
+				// 	c,
+				// 	c,
+				// 	"",
+				// )
+				//
+				// if err := api.Pin().Add(req.Context, rp, options.Pin.Recursive(true)); err != nil {
+
+				ret := RootMeta{Cid: &c, PresentInCar: seen}
+
+				if block, err := node.Blockstore.Get(c); err != nil {
+					ret.PinError = err
+				} else if nd, err := ipld.Decode(block); err != nil {
+					ret.PinError = err
+				} else if err := node.Pinning.Pin(req.Context, nd, true); err != nil {
+					ret.PinError = err
+				} else if err := node.Pinning.Flush(req.Context); err != nil {
+					ret.PinError = err
+				}
+
+				if err := res.Emit(&CarImportOutput{Root: ret}); err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, event *CarImportOutput) error {
+
+			silent, _ := req.Options[silentOptionName].(bool)
+			if silent {
+				return nil
+			}
+
+			// this is a progress display event
+			if event.Root.Cid == nil {
+				_, err := fmt.Fprintf(
+					w,
+					"Objects/DeclaredRoots/FoundRoots: %d/%d/%d\r",
+					event.ObjectCounts.Blocks,
+					event.ObjectCounts.RootsDeclared,
+					event.ObjectCounts.RootsSeen,
+				)
+				return err
+			}
+
+			enc, err := cmdenv.GetLowLevelCidEncoder(req)
+			if err != nil {
+				return err
+			}
+
+			var notInCar string
+			if !event.Root.PresentInCar {
+				notInCar = " (specified in header without its data)"
+			}
+
+			if event.Root.PinError != nil {
+				event.Root.PinError = fmt.Errorf("FAILED: %s", event.Root.PinError)
+			} else {
+				event.Root.PinError = errors.New("success")
+			}
+
+			_, err = fmt.Fprintf(
+				w,
+				"Pinned root %s%s: %s\n",
+				enc.Encode(*event.Root.Cid),
+				notInCar,
+				event.Root.PinError,
+			)
+			return err
+		}),
+	},
+}
+
+func importWorker(req *cmds.Request, re *cmds.ResponseEmitter, api *iface.CoreAPI, counts *ObjectCounts, ret chan importResult) {
+
+	// this is *not* a transaction
+	// it is simply a way to relieve pressure on the blockstore
+	// similar to pinner.Pin/pinner.Flush
+	batch := ipld.NewBatch(req.Context, (*api).Dag())
+
+	// It is not guaranteed that a root in a header is actually present in the same ( or any )
+	// .car file. This is the case in version 1, and ideally in further versions too
+	// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+	// We will attempt a pin *only* at the end in case all car files were well formed
+	//
+	// The boolean value indicates whether we have encountered the root within the car file's
+	expectedRootsSeen := make(expectedRootsSeen)
+
+	it := req.Files.Entries()
+	for it.Next() {
+
+		file := files.FileFromEntry(it)
+		if file == nil {
+			ret <- importResult{err: errors.New("expected a file handle")}
+			return
+		}
+		defer func() { file.Close() }()
```

Comment:
this will cause all the files to stay open until the end of the function. seems like worth having the file be closed within the scope - either factoring out the body into a function, or explicitly closing each loop iteration.

Reply:
As a "custom" we don't close anything in the CLI in the first place, this is kinda new. By design `it` opens everything there is to open ahead of time ( I left a comment about that). But yeah good point, defer removed!